addon/MetadataParser.js

```diff
@@ -0,0 +1,45 @@
+/* globals require, exports */
+"use strict";
+
+const {getMetadata} = require("common/vendor")("PageMetadataParser");
+const {Cc, Ci} = require("chrome");
+
+function MetadataParser(previewProvider) {
+  this._previewProvider = previewProvider;
+}
+
+MetadataParser.prototype = {
+  /**
+   * Parse the outerHTML
+   */
+  _getDocumentObject(text) {
+    const parser = Cc["@mozilla.org/xmlextras/domparser;1"]
+                  .createInstance(Ci.nsIDOMParser);
+    return parser.parseFromString(text, "text/html");
+  },
+
+  /**
+   * Format the data so the metadata DB receives it
+   */
+  _formatData(data, url) {
+    data.images = data.image_url ? [{url: data.image_url, height: 500, width: 500}] : [];
+    data.favicon_url = data.icon_url ? data.icon_url : "";
+    data.url = url;
+    delete data.image_url;
+    delete data.icon_url;
+    let link = this._previewProvider.processLinks([data]);
```

Comment:
Let's not do this here at all.  If we simply pre process the link into an object with the fields that preview provider will accept then we can remove our dependency on it here.